src/v/cluster/topic_table.cc

```diff
@@ -647,28 +649,42 @@ void topic_table::notify_waiters() {
     /// If by invocation of this method there are no waiters, notify
     /// function_ptrs stored in \ref notifications, without consuming all
     /// pending_deltas for when a subsequent waiter does arrive
-    std::vector<delta> changes;
-    std::copy_if(
-      _pending_deltas.begin(),
-      _pending_deltas.end(),
-      std::back_inserter(changes),
-      [this](const delta& d) { return d.offset > _last_consumed_by_notifier; });
-    for (auto& cb : _notifications) {
-        cb.second(changes);
-    }
+
+    // \ref notify_waiters is called after every apply. Hence for the most
+    // part there should only be a few items towards the end of \ref
+    // _pending_deltas that need to be sent to callbacks in \ref
+    // notifications. \ref _last_consumed_by_notifier_offset allows us to
+    // skip ahead to the items added since the last \ref notify_waiters
+    // call.
+    auto starting_iter = _pending_deltas.cbegin()
+                         + _last_consumed_by_notifier_offset;
+
+    std::span changes{starting_iter, _pending_deltas.cend()};
+
     if (!changes.empty()) {
+        for (auto& cb : _notifications) {
+            cb.second(changes);
+        }
+
         _last_consumed_by_notifier = changes.back().offset;
```

Comment:
it's currently used in 2 places `src/v/coproc/reconciliation_backend.cc` and `src/v/kafka/server/group_manager.cc`.

Reply:
It has one current use [here](https://github.com/redpanda-data/redpanda/blob/e58aa00950ee751af624e6b01b83a3f4b515026a/src/v/cluster/topic_table.cc#L798). It may be able to be replaced with `_last_applied_revision_id`. However, in some cases `_last_applied_revision_id` is greater than `_last_consumed_by_notifier` since `_last_consumed_by_notifier` won't be updated if an apply function returns an error but `_last_applied_revision_id` will be. So just to be safe I'll leave it in there for now.

Reply:
Hmmm, OK. That line honestly looks like a bug. It is comparing `model::revision` with a `model::offset` which I think does not ever make sense.

It's a shame because I feel like things are getting complicated: we now have `_last_consumed_by_notifier` and `_last_consumed_by_notifier_offset` yet the former is actually an _offset_ (in the log sense) and the latter is an offset into the pending deltas, and they serve basically the same purpose.

Reply:
Nevermind I guess it is not a bug: it is valid to compare the revision ID to the offset in this case (see slack). My uneasiness remains regarding these two very similar bits of state but I guess it is what it is.